### PR TITLE
Redact request url

### PR DESCRIPTION
**Pull request by tgwizard**
## Goal

Currently, when the Bugsnag client extracts the URL from the request, it includes the full query parameters, without any filtering/redaction. This will leak any sensitive keys sent as query parameters, like access tokens.

This PR changes that so that we go through each query parameter and filter out sensitive ones (similar to how we do for request headers).

This PR also adds `access_token` as a default sensitive parameter, since it's very unlikely users want to include those in Bugsnag reports.

### Changed

`extractRequestInfoFromReq`

## Tests

I've added tests for the function.

### Outstanding Questions

One drawback of this change is that the URL query parameters sent to Bugsnag will no longer be in the order that the client sent them, but sorted alphabetically by key. This is because `url.ParseQuery` returns a (type-aliased) `map[string][]string`, and the corresponding `Encode` sorts on key.

We could attempt a custom implementation of the `Encode` function to sort according to the original order, but I'm not sure it's worth it.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [x] Consistency across platforms for structures or concepts added or modified
- [x] Consistency between the changeset and the goal stated above
- [x] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [x] Usage friction - is the proposed change in usage cumbersome or complicated?
- [x] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [x] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [x] Thoroughness of added tests and any missing edge cases
- [x] Idiomatic use of the language

### CODE REVIEW

**Comment by tgwizard** [comment]
Please review :)

File: request_extractor.go [hunk 1]
```diff
@@ -32,14 +33,39 @@ func extractRequestInfo(ctx context.Context) (*RequestJSON, *http.Request) {
...
+
+	var rawQuery string
+	parsedQuery, err := url.ParseQuery(req.URL.RawQuery)
+	if err != nil {
+		rawQuery = req.URL.RawQuery
+	} else {
+		for key, values := range parsedQuery {
+			if contains(Config.ParamsFilters, key) {
+				for i, v := range values {
+					if len(v) != 0 {
```
[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor.go]: Yep, exactly.

File: request_extractor.go [hunk 2]
```diff
@@ -32,14 +33,39 @@ func extractRequestInfo(ctx context.Context) (*RequestJSON, *http.Request) {
...
-	proto := "http://"
+	scheme := "http"
 	if req.TLS != nil {
-		proto = "https://"
+		scheme = "https"
 	}
+
+	var rawQuery string
+	parsedQuery, err := url.ParseQuery(req.URL.RawQuery)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor.go]: It can, I added a test case.

File: request_extractor.go [hunk 3]
```diff
@@ -32,14 +33,38 @@ func extractRequestInfo(ctx context.Context) (*RequestJSON, *http.Request) {
...
+		rawQuery = req.URL.RawQuery
+	} else {
+		for key, values := range parsedQuery {
+			if contains(Config.ParamsFilters, key) {
+				for i, v := range values {
+					if len(v) != 0 {
+						values[i] = "FILTERED"
```
[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor.go]: The reason I skipped the `[` and `]` wrappers is that they are URL-encoded when building the URL, so `/foo.html?access_token=something` becomes `/foo.html?access_token=%5BFILTERED%5D`. I think that's more difficult to read.

I can change to `[FILTERED]` if you want still though.

File: request_extractor_test.go [hunk 4]
```diff
@@ -54,6 +55,35 @@ func TestRequestExtractorCanHandleAbsentContext(t *testing.T) {
...
+		{"/foo.html?q=something&bar=123", "http://example.com/foo.html?bar=123&q=something"},
+		{"/foo.html?foo=1&foo=2&foo=3", "http://example.com/foo.html?foo=1&foo=2&foo=3"},
+
+		{"/foo.html?access_token=something", "http://example.com/foo.html?access_token=FILTERED"},
+		{"/foo.html?access_token=something&access_token=", "http://example.com/foo.html?access_token=FILTERED&access_token="},
+	}
+
+	for _, tc := range testCases {
+		parsedURL, err := url.Parse(tc.originalURI)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor_test.go]: I kept this but updated the wording - it's to make sure that you write correct input URLs.

**Comment by tgwizard** [comment]
Thanks for the review @kinbiko! I've pushed a commit addressing most of your comments, please have another look. I'd be happy to squash the commits at merge time.

**Comment by tgwizard** [comment]
@kinbiko please review again :)

File: request_extractor.go [hunk 5]
```diff
@@ -32,19 +33,55 @@ func extractRequestInfo(ctx context.Context) (*RequestJSON, *http.Request) {
...
+		changed := false
+		for key, values := range parsedQuery {
+			if contains(Config.ParamsFilters, key) {
+				for i, v := range values {
+					if len(v) == 0 {
+						// No need to filter empty parameters.
+						continue
+					}
+
+					values[i] = "FILTERED"
```
Comment (kinbiko) [inline request_extractor.go]: Good point about `[]` being represented differently in a URL, I didn't think of that. I think we can get around this though by doing a string replacement after we've rendered the URL.
Just in case `"FILTERED"` is ever used as a (valid) URL param value, maybe we should do
```suggestion
					values[i] = "BUGSNAG_URL_FILTERED"
```
here, and then replace `"BUGSNAG_URL_FILTERED"` with `"[FILTERED]"` just before returning from this function.

File: request_extractor.go [hunk 6]
```diff
@@ -32,19 +33,55 @@ func extractRequestInfo(ctx context.Context) (*RequestJSON, *http.Request) {
...
+// The query part of the URL might appear differently (different order of parameters) if any filtering was done.
+func sanitizeURL(req *http.Request) string {
+	scheme := "http"
+	if req.TLS != nil {
+		scheme = "https"
+	}
+
+	rawQuery := req.URL.RawQuery
+	parsedQuery, err := url.ParseQuery(req.URL.RawQuery)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor.go]: I can do that, but to return early I need too copy the URL building code (lines 77-83) into the `if err != nil` clause. I think this is clearer, with less duplication. Do you still think I should return early?

**Comment by tgwizard** [comment]
I've addressed all but one of your comments, which I commented on why. Please have another look @kinbiko.

File: request_extractor.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kinbiko) [reply request_extractor.go]: I don't see why we need to build the URL again, but I think maybe I'm missing something here. 
Here's my thinking: We wish to return the URL string, which we have all the components for presumably, i.e. `scheme`, some forward slashes, `req.Host`, and `req.RequestURI`. Can we not concatenate these in a one-liner if `err != nil`?

[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor.go]: My reason to always use the same code path is if we want to include any `user:[FILTERED]` user/password in the extracted URL, or to simplify if we want more processing (and to only have one code-path that builds the URL).

But I've pushed a commit to change to what you want.

[the remark this answers is not in the document]
Comment (tgwizard) [reply request_extractor.go]: Also, we're not really building the URL again - we don't have the URL, we're building it for the first time here.

**Comment by tgwizard** [comment]
The tests fail for some seemingly unrelated reason - a timeout in the Martini tests.

**Comment by tgwizard** [comment]
@kinbiko please have another look, it should now all be fixed.

**Comment by kinbiko** [comment]
Hi @tgwizard,

Sorry about the wait -- I was struggling to find the time to come back and re-review this. I've taken this branch for a spin now, and it looks really good; only one minor issue.

Unfortunately, not filtering empty URL params leads to a pretty confusing Request tab; The URL displays an empty value but the `params` extracted from the built-in Bugsnag HTTP middleware show up as `"[FILTERED]"`.

image

I've had a chat with another of the maintainers here and we believe it's best to be consistent and use `"[FILTERED]"` even for empty fields.

That being said, we do see the benefit of being able to tell if such fields are absent. If this is something you need for your particular use case you may use an [`OnBeforeNotify`](https://docs.bugsnag.com/platforms/go/net-http/customizing-error-reports/#modifying-report-information) callback to extract this information, so all is not lost 🙂.

Other than this, I believe this change is good to go. Thank you ever so much for your patience. I realise I've been making you jump through a lot of hoops.

You are free to address this if you so wish, otherwise I'll merge and address myself as soon as I find the time.

**Comment by tgwizard** [comment]
No worries @kinbiko, I'm happy to make this right. The code is much better now than in my initial PR, which also makes me happy 😸 .

I've added a commit to also add `[FILTERED]` for empty values for sensitive query string parameters.

**Comment by tgwizard** [comment]
@kinbiko bump 😺

**Comment by snmaynard** [comment]
@tgwizard I believe @kinbiko is fixing up our CI issues on older Go versions and then the plan is for this to be merged. Should be done in the next day or so I think (but correct me if I'm wrong @kinbiko!)

**Comment by kinbiko** [comment]
Just as @snmaynard mentioned, we were having some CI issues that I needed to fix before this could go in. This is why your latest build failed (nothing to do with your changes), but the PR looks good now so it's going in our next release (most likely shipping Monday). 🚢

Thanks again for your work on this @tgwizard!
